include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h

```diff
@@ -3831,6 +3830,491 @@ MGTwoLevelTransfer<dim, VectorType>::memory_consumption() const
+          evaluator.distribute_local_to_global_unconstrained(weight_vector);
+        }
+      weight_vector.compress(VectorOperation::add);
+
+      for (unsigned int i = 0; i < weight_vector.locally_owned_size(); ++i)
+        if (weight_vector.local_element(i) > 0.)
+          weight_vector.local_element(i) =
+            Number(1.) / weight_vector.local_element(i);
+      for (const unsigned int index :
+           matrix_free_fine.get_constrained_dofs(dof_no_fine))
+        weight_vector.local_element(index) = 0;
+      weight_vector.update_ghost_values();
+
+      weights.clear();
+      weights.reserve(matrix_free_fine.n_cell_batches() *
+                      Utilities::pow(3, dim));
+      weights_start.clear();
+      weights_start.resize(matrix_free_fine.n_cell_batches());
+      std::array<VectorizedArrayType, Utilities::pow(3, dim)>
+        weights_compressed;
+      for (unsigned int cell = 0; cell < matrix_free_fine.n_cell_batches();
+           ++cell)
+        {
+          weights_start[cell] = weights.size();
+
+          weights_compressed.fill(VectorizedArray<Number>());
+          evaluator.reinit(cell);
+          evaluator.read_dof_values_unconstrained(weight_vector);
+
+          const bool can_compress_weights =
+            internal::compute_weights_fe_q_dofs_by_entity<dim, -1>(
+              evaluator.begin_dof_values(),
+              1,
+              fe_fine.degree + 1,
+              weights_compressed.data());
+          if (can_compress_weights && fe_fine.degree > 1)
+            for (const VectorizedArrayType weight : weights_compressed)
+              weights.push_back(weight);
+          else
+            for (unsigned int i = 0; i < evaluator.dofs_per_cell; ++i)
+              weights.push_back(evaluator.begin_dof_values()[i]);
+        }
+      weights_start.push_back(weights.size());
+    }
+
+  // Compute interpolation matrices, possibly in the 1D version
+  const FiniteElement<dim> &fe_coarse = dof_coarse.get_fe();
+
+  const auto reference_cell = fe_fine.reference_cell();
+
+  Assert(reference_cell == fe_coarse.reference_cell(), ExcNotImplemented());
+
+  if (reference_cell == ReferenceCells::get_hypercube<dim>())
+    {
+      const std::unique_ptr<FiniteElement<1>> fe_fine_1d =
+        internal::create_1D_fe(fe_fine);
+      std::vector<unsigned int> renumbering_fine(fe_fine_1d->n_dofs_per_cell());
+      {
+        AssertIndexRange(fe_fine_1d->n_dofs_per_vertex(), 2);
+        renumbering_fine[0] = 0;
+        for (unsigned int i = 0; i < fe_fine_1d->dofs_per_line; ++i)
+          renumbering_fine[i + fe_fine_1d->n_dofs_per_vertex()] =
+            GeometryInfo<1>::vertices_per_cell *
+              fe_fine_1d->n_dofs_per_vertex() +
+            i;
+        if (fe_fine_1d->n_dofs_per_vertex() > 0)
+          renumbering_fine[fe_fine_1d->n_dofs_per_cell() -
+                           fe_fine_1d->n_dofs_per_vertex()] =
+            fe_fine_1d->n_dofs_per_vertex();
+      }
```

Comment:
All of this part (and some 90 lines below) repeat more or less what we did for the other transfer. Also, there is some repetition for the weights. If this PR finds approval, I will unify the code (either here or in a separate PR), but I wanted to get the general concept approved before I start refactoring.

Reply:
I now pushed a new version that unifies the two places, making the added code much nicer.